tests/test_structured.cpp

```diff
@@ -0,0 +1,239 @@
+    test_numeric_to_string<unsigned long long>();
+    test_numeric_to_string<unsigned char>();
+    test_numeric_to_string<float>();
+    test_numeric_to_string<double>();
+    test_numeric_to_string<long double>();
+
+    // string_view
+    REQUIRE(spdlog::details::value_to_string(spdlog::Field("", "")) == "");
+    REQUIRE(spdlog::details::value_to_string(spdlog::Field("", "data")) == "data");
+
+    // bool
+    REQUIRE(spdlog::details::value_to_string(spdlog::Field("", true)) == "true");
+    REQUIRE(spdlog::details::value_to_string(spdlog::Field("", false)) == "false");
+
+    // char
+    REQUIRE(spdlog::details::value_to_string(spdlog::Field("", 'c')) == "c");
+
+}
+
+
+TEST_CASE("buffered_msg_field_copies ", "[structured]")
+{
+    std::unique_ptr<spdlog::details::log_msg_buffer> test1;
+    {
+        auto array2 = {F("var", 1), F("var2", "two")};
+        spdlog::details::log_msg test_input1(spdlog::source_loc{}, "name", spdlog::level::info, "msg", array2.begin(), array2.size());
+        test1 = spdlog::details::make_unique<spdlog::details::log_msg_buffer>(test_input1);
+    }
+    REQUIRE(test1->field_data_count == 2);
+    REQUIRE(test1->field_data[0].name == "var");
+    REQUIRE(test1->field_data[0].value_type == spdlog::FieldValueType::INT);
+    REQUIRE(test1->field_data[0].int_ == 1);
+    REQUIRE(test1->field_data[1].name == "var2");
+    REQUIRE(test1->field_data[1].value_type == spdlog::FieldValueType::STRING_VIEW);
+    REQUIRE(test1->field_data[1].string_view_ == "two");
+}
+
+TEST_CASE("async_structured ", "[structured]")
+{
+    auto test_sink = std::make_shared<spdlog::sinks::test_sink_mt>();
+    test_sink->set_delay(std::chrono::milliseconds(1));
+    size_t queue_size = 4;
+    size_t messages = 1024;
+
+    auto tp = std::make_shared<spdlog::details::thread_pool>(queue_size, 1);
+    auto logger = std::make_shared<spdlog::async_logger>("as", test_sink, tp, spdlog::async_overflow_policy::overrun_oldest);
+    for (size_t i = 0; i < messages; i++)
+    {
+        // Build on the stack so it's out of scope
+        std::string test_string("abcdefghijklmnopqrstuvwxyz");
+        logger->log({}, spdlog::level::info, {F("str", test_string)}, "test msg");
+    }
+    logger->flush();
+    REQUIRE(test_sink->msg_counter() < messages);
+    REQUIRE(test_sink->msg_counter() > 0);
+    REQUIRE(tp->overrun_counter() > 0);
+}
+
+
+#if SPDLOG_ACTIVE_LEVEL != SPDLOG_LEVEL_DEBUG
+#    error "Invalid SPDLOG_ACTIVE_LEVEL in test. Should be SPDLOG_LEVEL_DEBUG"
+#endif
+
+#define TEST_FILENAME "test_logs/structured_macro_log"
+std::string last_line(const std::string &str) {
+    if (str.empty()) {
+        return "";
+    }
+    auto start = str.begin();
+    auto curr = str.end() - 1;
+
+    // Skip last line
+    while ((*curr == '\n' || *curr == '\r') && curr != start) {
+        curr--;
+    }
+    auto end = curr + 1; // end is exclusive
+
+    while (curr != start && *(curr-1) != '\n' && *(curr-1) != '\r' && curr != start) {
```

Comment:
Inadvertent duplication?

```suggestion
    while (curr != start && *(curr-1) != '\n' && *(curr-1) != '\r') {
```

Reply:
I wanted to really, really make sure that we weren't at `start`.  :-)